src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs

```diff
@@ -298,21 +300,35 @@ public Task<IList<Message>> ReceiveAsync(int maxMessageCount)
 
             MessagingEventSource.Log.MessageReceiveStart(this.ClientId, maxMessageCount);
 
+            bool isDiagnosticsEnabled = ServiceBusDiagnosticSource.IsEnabled();
+            Activity activity = isDiagnosticsEnabled ? this.diagnosticSource.ReceiveStart(maxMessageCount) : null;
+            Task receiveTask = null;
+
             IList<Message> unprocessedMessageList = null;
             try
             {
-                await this.RetryPolicy.RunOperation(
+                receiveTask = this.RetryPolicy.RunOperation(
                     async () =>
                     {
-                        unprocessedMessageList = await this.OnReceiveAsync(maxMessageCount, operationTimeout).ConfigureAwait(false);
-                    }, operationTimeout)
-                    .ConfigureAwait(false);
+                        unprocessedMessageList = await this.OnReceiveAsync(maxMessageCount, operationTimeout)
+                            .ConfigureAwait(false);
+                    }, operationTimeout);
+                await receiveTask.ConfigureAwait(false);
+
             }
             catch (Exception exception)
             {
                 MessagingEventSource.Log.MessageReceiveException(this.ClientId, exception);
+                if (isDiagnosticsEnabled)
+                {
+                    this.diagnosticSource.ReportException(exception);
```

Comment:
Consider putting this before the call to `MessagingEventSource.Log.MessageReceiveException`. That way, I think, we will have a consistent nesting of EventSource (outer) and DiagnosticSource (inner) messages.

Reply:
makes sence